controllers/topology/internal/extensions/patches/patch.go

```diff
@@ -0,0 +1,198 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package patches
+
+import (
+	"context"
+	"strings"
+
+	"github.com/pkg/errors"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/contract"
+	tlog "sigs.k8s.io/cluster-api/controllers/topology/internal/log"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/mergepatch"
+)
+
+// PatchOption represents an option for the patchObject and patchTemplate funcs.
+type PatchOption interface {
+	// ApplyToHelper applies configuration to the given options.
+	ApplyToHelper(*PatchOptions)
+}
+
+// PatchOptions contains options for patchObject and patchTemplate.
+type PatchOptions struct {
+	preserveFields []contract.Path
+}
+
+// ApplyOptions applies the given patch options
+// and then returns itself (for convenient chaining).
+func (o *PatchOptions) ApplyOptions(opts []PatchOption) *PatchOptions {
+	for _, opt := range opts {
+		opt.ApplyToHelper(o)
+	}
+	return o
+}
+
+// PreserveFields instructs the patch func to preserve fields.
+type PreserveFields []contract.Path
+
+// ApplyToHelper applies this configuration to the given patch options.
+func (i PreserveFields) ApplyToHelper(opts *PatchOptions) {
+	opts.preserveFields = i
+}
+
+// patchObject overwrites spec in object with spec.template.spec of patchedTemplate,
+// while preserving the configured fields.
+func patchObject(ctx context.Context, object, patchedTemplate *unstructured.Unstructured, opts ...PatchOption) error {
+	log := tlog.LoggerFrom(ctx)
+
+	patchOptions := &PatchOptions{}
+	patchOptions = patchOptions.ApplyOptions(opts)
+
+	// Create a copy of the object.
+	original := object.DeepCopy()
+
+	if err := copySpec(copySpecInput{
+		src:              patchedTemplate,
+		dest:             object,
+		srcSpecPath:      "spec.template.spec",
+		destSpecPath:     "spec",
+		fieldsToPreserve: patchOptions.preserveFields,
+	}); err != nil {
+		return errors.Wrapf(err, "failed to patch object %s by appying changes from the patched template",
+			tlog.KObj{Obj: object})
+	}
+
+	// Log the delta between the object before and after applying the accumulated patches.
+	helper, err := mergepatch.NewHelper(original, object, nil)
```

Comment:
I'm not sure it is appropriate using mergepatch.NewHelper to track the difference between objects, given that it computes a two way patch which preserve values from the original object, while instead we are overriding the entire spec (same for patch template)

Reply:
100% agree, it's not appropriate. I'll add a small diff util func

Reply:
Fixed, PTAL